proxylib/cassandra/cassandraparser.go

```
@@ -0,0 +1,643 @@
+			}
+
+		}
+
+		log.Debugf("Parsed CassandraRule pair: %v", cr)
+		rules = append(rules, &cr)
+	}
+	return rules
+}
+
+type CassandraParserFactory struct{}
+
+var cassandraParserFactory *CassandraParserFactory
+
+func init() {
+	log.Info("init(): Registering cassandraParserFactory")
+	RegisterParserFactory("cassandra", cassandraParserFactory)
+	RegisterL7RuleParser("cassandra", CassandraRuleParser)
+}
+
+type CassandraParser struct {
+	connection *Connection
+	inserted   bool
+	keyspace   string // stores current keyspace name from 'use' command
+
+	// stores prepared query string while
+	// waiting for 'prepared' reply from server
+	// with a prepared id.
+	// replies associated via stream-id
+	preparedQueryPathByStreamID map[uint16]string
+
+	// allowing us to enforce policy on query
+	// at the time of the execute command.
+	preparedQueryPathByPreparedID map[string]string // stores query string based on prepared-id,
+}
+
+func (pf *CassandraParserFactory) Create(connection *Connection) Parser {
+	log.Debugf("CassandraParserFactory: Create: %v", connection)
+
+	p := CassandraParser{connection: connection}
+	p.preparedQueryPathByStreamID = make(map[uint16]string)
+	p.preparedQueryPathByPreparedID = make(map[string]string)
+	return &p
+}
+
+func (p *CassandraParser) OnData(reply, endStream bool, dataArray [][]byte, offset int) (OpType, int) {
+
+	// inefficient, but simple for now
+	data := bytes.Join(dataArray, []byte(""))[offset:]
+	log.Debugf("OnData offset  = %d", offset)
+
+	if len(data) < cassHdrLen {
+		// Partial header received, ask for more
+		needs := cassHdrLen - len(data)
+		log.Debugf("Did not receive full header, need %d more bytes", needs)
+		return MORE, needs
+	}
+
+	// full header available, read full request length
+	requestLen := binary.BigEndian.Uint32(data[5:9])
+	log.Debugf("Request length = %d", requestLen)
+	if requestLen > cassMaxLen {
+		log.Errorf("Request length of %d is greater than 256 MB", requestLen)
+		return ERROR, 0
```

Should return the defined error code here, like:
```
	return ERROR, int(ERROR_INVALID_FRAME_LENGTH)
```